### PR TITLE
feat: Draft fledgeling object, ready for internal use

### LINKED ISSUE #402: Simple fledgling object
Part of #331.

This object would only parse top-level headers (version, date, and perhaps format of the header), and contain the contents of the Markdown chunks between top-level headers as character vectors. Also, the relevant parts of `DESCRIPTION` (current version and date).

No tinkr. Parsing with regular expressions (e.g. `"^ *# +..."`) .

As simple as possible, to get us started.

### PR DESCRIPTION
**Pull request by krlmlr**
Closes #402.

TODOS

- [x] Also parse h2 for version (see pkgdown)
- [x] tests

### CODE REVIEW

**Comment by krlmlr** [comment]
@maelle: Can you please pick up from here?

File: R/fledgeling.R [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  package <- unname(desc::desc_get("Package"))
+  version <- desc::desc_get_version()
+
+  if (file.exists("NEWS.md")) {
+    news <- readLines("NEWS.md")
+  } else {
+    news <- character()
+  }
+
+  header_rx <- "^# +[a-zA-Z][a-zA-Z0-9.]*[a-zA-Z0-9] +(?<version>[0-9][0-9.-]*) *(?<date>.*)$"
```
Comment (maelle) [inline R/fledgeling.R:31]: note to self: to be tried with version nicknames

File: R/fledgeling.R [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+  first_level_headers <- grep(header_rx, news, perl = TRUE)
```
Comment (maelle) [inline R/fledgeling.R:33]: I still think it'd be easier with tinkr :innocent: but understand the need for not using too many deps

File: NEWS.md [hunk 3]
```diff
@@ -1,13 +1,13 @@
 <!-- NEWS.md is maintained by https://cynkra.github.io/fledge, do not edit -->
 
-# fledge 0.1.0.9022 (2022-11-10)
+# fledge 0.1.0.9022 (2022-11-10) "blabla blop"
```
Comment (maelle) [inline NEWS.md:3]: to be replaced with actual tests asap :-)

**Comment by maelle** [comment]
@krlmlr I added a bit more parsing, and tests, you can now review this.

**Comment by maelle** [comment]
Working on writing should happen after both this and #584 are merged, but I'm working on paper in the meantime.

I feel parsing the news a bit more should happen now too.

File: R/fledgeling.R [hunk 4]
```diff
@@ -0,0 +1,79 @@
+#' fledgeling object
```
Comment (maelle) [inline R/fledgeling.R:1]: why not "fledgling" (removing the second e)

Comment (krlmlr) [reply R/fledgeling.R:1]: No strong preference.

**Comment by maelle** [comment]
why is the object immutable, more specifically, why not copy more of desc logic?

**Review by krlmlr — COMMENTED** [review]
Thanks. The h2 detection is broken, but we also don't need it here. Can you please remove it? Parsing the individual h1 sections would be done on demand, later.

**Comment by krlmlr** [comment]
Can you share an example NEWS file where the h2 detection is triggered?

**Comment by krlmlr** [comment]
The `original` column should perhaps contain the entire section, including empty lines. Perhaps as a single character string, with embedded `"\n"` ?

**Comment by maelle** [comment]
why would the header and content be stored together?

File: R/fledgeling.R (hunk 4 above)
Comment (maelle) [reply R/fledgeling.R:1]: will rename

**Comment by krlmlr** [comment]
I have code that also extracts the preamble: https://github.com/cynkra/fledge/tree/f-402-fledgeling-3 .

**Comment by maelle** [comment]
> Should the class name be "fledge_ling" ? So that the package name still is the prefix?

but what about method names?

**Comment by krlmlr** [comment]
To avoid conflicts with your work. Pushing now.

**Comment by krlmlr** [comment]
Also keeping raw text now for preamble and all sections. We can do a perfect roundtrip with this object now, and should start using it everywhere in the codebase.

**Comment by krlmlr** [comment]
Fixed corner case with empty preamble.

Ready for review, talk soon!

**Review by maelle — APPROVED** [review]
should this be merged?

**Pull request merged**